firmware/src/dap.rs

```diff
@@ -472,40 +472,68 @@ impl<'a> DAP<'a> {
         let mask = req.next_u8();
         let wait = req.next_u32();
 
-        // Our pin mapping:
-        // SWDIO/TMS: FLASH_SI
-        // SWCLK/TCK: SCK
-        // SWO/TDO: FLASH_CS
-        // TDI: FPGA_RST
-        // nRESET: FLASH_SO
-        //
         // SWJ_Pins mapping:
         // 0: SWCLK/TCK
         // 1: SWDIO/TMS
         // 2: TDI
         // 3: TDO
         // 5: nTRST
         // 7: nRESET
-        //
-        // We only support setting nRESET.
 
         const SWCLK_POS: u8 = 0;
         const SWDIO_POS: u8 = 1;
         const TDI_POS: u8 = 2;
         const TDO_POS: u8 = 3;
         const NTRST_POS: u8 = 5;
         const NRESET_POS: u8 = 7;
-        const NRESET_MASK: u8 = 1 << NRESET_POS;
 
-        // If reset bit is in mask, apply output bit to pin
-        if (mask & NRESET_MASK) != 0 {
-            if output & NRESET_MASK != 0 {
-                self.pins.reset.set_high();
-            } else {
-                self.pins.reset.set_low();
-            }
+        macro_rules! set_pins {
+            (
+                $($bit_position:ident: $pin:ident $({ $mode:ident })?, )+
+            ) => {{
+                $(
+                    if (mask & (1 << $bit_position)) != 0 {
+                        $( self.pins.$pin.$mode(); )?
+                        if output & (1 << $bit_position) != 0 {
+                            self.pins.$pin.set_high();
+                        } else {
+                            self.pins.$pin.set_low();
+                        }
+                    }
+                )+
+            }};
         }
 
+        match self.mode {
+            Some(DAPMode::SWD) => {
+                // SWD on idle runs GPIO in alternate mode (SPI peripheral), mode change
+                // required
+                set_pins!(
+                    SWDIO_POS: spi1_mosi { set_mode_output },
+                    SWCLK_POS: spi1_clk { set_mode_output },
+                );
+            }
+            Some(DAPMode::JTAG) => {
+                // JTAG on idle runs GPIO in bitbang mode, no mode changes required
+                //
+                // TDO in an input pin, ignored
+                // Note: DAPLink implementation also ignores TDO pin; even though CMSIS-DAP's
+                // SWJ_Pins command should allow to control it
+                set_pins!(
+                    SWDIO_POS: spi1_mosi,
+                    SWCLK_POS: spi2_clk,
+                    TDI_POS: spi2_mosi,
+                );
+            }
+            None => {
+                resp.write_err();
```

Comment:
Ah, actually, this is wrong in a misleading way: the SWJ_Pins command does not have a status code, it always returns the state of the pins. Calling `write_err()` returns 0xFF instead, which the probe wants to mean "an error occurred" but host software interprets to mean "all pins are high".

I think this is what was masking a problem with `--connect-under-reset`, which indeed was sending SWJ_Pins _before_ calling connect. Since it doesn't expect an error it continued and eventually failed to program (though I'm not sure what caused the failure since it works OK without `--connect-under-reset`...)

Reply:
Replacing the `None` branch with an empty block (so that nReset is still processed when not in any mode) fixes the `--connect-under-reset` issue (and I think makes sense as the thing to do).

Possibly we should also respect other pins when not yet connected too though, it seems like some applications might want to run their special sequence before connecting?